raiden_contracts/contracts/TokenNetwork.sol

```diff
@@ -329,78 +329,88 @@ contract TokenNetwork is Utils {
     /// amount.
     /// @param total_withdraw Total amount of tokens that are marked as
     /// withdrawn from the channel during the channel lifecycle.
-    /// @param partner Channel partner address, needed to compute the total
-    /// channel deposit.
     /// @param participant_signature Participant's signature on the withdraw
     /// data.
     /// @param partner_signature Partner's signature on the withdraw data.
     function setTotalWithdraw(
         uint256 channel_identifier,
         address participant,
         uint256 total_withdraw,
-        address partner,
         bytes participant_signature,
         bytes partner_signature
     )
         isOpen(channel_identifier)
         external
     {
-        require(channel_identifier == getChannelIdentifier(participant, partner));
+        uint256 total_deposit;
+        uint256 current_withdraw;
+        address partner;
+
         require(total_withdraw > 0);
 
-        verifyWithdrawSignatures(
+        // Authenticate both channel partners via there signatures:
+        require(participant == recoverAddressFromWithdrawMessage(
+            channel_identifier,
+            participant,
+            total_withdraw,
+            participant_signature
+        ));
+        partner = recoverAddressFromWithdrawMessage(
             channel_identifier,
             participant,
-            partner,
             total_withdraw,
-            participant_signature,
             partner_signature
         );
 
-        uint256 total_deposit;
-        uint256 current_withdraw;
+        // Validate that authenticated partners and the channel identifier match
+        require(channel_identifier == getChannelIdentifier(participant, partner));
 
+        // Read channel state after validating the function input
         Channel storage channel = channels[channel_identifier];
         Participant storage participant_state = channel.participants[participant];
         Participant storage partner_state = channel.participants[partner];
 
         total_deposit = participant_state.deposit + partner_state.deposit;
-        // This should never happen, as we have an overflow check in setTotalDeposit
-        assert(total_deposit >= participant_state.deposit);
-        assert(total_deposit >= partner_state.deposit);
 
         // Entire withdrawn amount must not be bigger than the current channel deposit
-        require(total_withdraw <= (total_deposit - partner_state.withdrawn_amount));
+        require((total_withdraw + partner_state.withdrawn_amount) <= total_deposit);
```

Comment:
Overflows are out of the agreed upon contracts scope here (i.e. assuming one misbehaving client max.) - signatures from both clients are required, so they would have to agree upon a corrupted withdrawal request, where one party would loose tokens. I can add a check anyhow.

Reply:
> signatures from both clients are required, so they would have to agree upon a corrupted withdrawal request

Correct. However, the overflow check that I mentioned makes sure that a participant (with a collaborating partner) cannot withdraw more tokens from the `TokenNetwork` contract.

Issue: `total_deposit` is big enough to cause an overflow on  `total_withdraw + partner_state.withdrawn_amount` & pass the check here: `require((total_withdraw + partner_state.withdrawn_amount) <= total_deposit);`

No other checks fail before `require(token.transfer(participant, current_withdraw));`. This can potentially transfer a lot of tokens to `participant` if the `TokenNetwork` contract has enough tokens or if the token contract is not properly implemented.